Add command reference docs

riyazdf opened the pull request:
This adds a command reference doc, so we can give our users a quicker way to find how to do things like rotate keys with notary CLI.

Inspired by and superseding #904 (thank you again @avaid96 for your work there!)

Signed-off-by: Riyaz Faizullabhoy <riyaz.faizullabhoy@docker.com>

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,235 @@
+## Terminology Reference
+1. **GUN**: Notary uses Globally Unique Names (GUNs) to identify trusted collections.
+2. **Target**: Notary refers to the files the framework is used to distribute as "target files".  Target files are opaque to the framework. Whether target files are packages containing multiple files, single text files, or executable binaries is irrelevant to Notary.
+3. **Trusted Collection**: A trusted collection is a conceptual source of target files of interest to the application.
```
I think I see what you mean by "source", but it may not be very clear since since notary doesn't necessarily identify the source of anything, and technically those files can be in disparate locations.  Maybe just "a list of target files"?  :|  I don't have a good suggestion.

cyli replied:
Actually maybe I don't really know what a trusted collection means.  I interpreted it to be a collection of the targets in a notary repo, and maybe that's not what it actually means.  If it is, would it make sense to add `notary repository` here as a term and say how it's different than the trusted collection it represents?

riyazdf replied:
I think `a list of target files` makes more sense, I'll add that and `notary repository` as well 👍

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,235 @@
+$ notary init <GUN> --rootkey <key_file>
+```
+
+Note that you will have to run a publish after this command, as there are staged changes to initialize the trusted collection that have not been pushed to a ntoary server:
````
nitpick: "notary"

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,235 @@
+$ notary key rotate <GUN> <key_role>
+```
+
+The Notary CLI client generates a new key for the role you specified, and
````
Would it make sense in this section to mention that `key rotate` replaces all existing keys on that role?

riyazdf replied:
Definitely 👍

cyli commented:
Would it make sense to add the add target/remove target references in this document too?  It's not very useful for docker users, but would be useful for manually signing things.  

Also, it might be useful to discuss the concept of staged changes and why notary does it and which types of commands produce staged changes.  I'm not sure if this should go into the  "manage staged changes" section or in other sections though, which is why I am adding an overall comment instead of a comment in a particular location.  :)

riyazdf commented:
@cyli agreed to both -- I initially left them out since they seemed basic enough from the getting started docs, but I think it'd be good to make this reference comprehensive.  I've also switched around the order of the first few sections and added a note about the "rebasing" notary does with local changelist files 👍

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,279 @@
+keys, authorize other team members to sign content, and rotate the keys if
+a private key has been compromised.
+
+When using the Notary CLI client you need to specify where is Notary server
```
grammar nit:  maybe "the URL of the Notary server..."?

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+$ notary init <GUN>
+```
+
+This command will generate targets and snapshot keys locally for the trusted collection, and try to locate a root key to use in the specified metadata cache (from `-d` or the config).
````
nit:  "notary client cache", since the keys are separate from the metadata cache specifically.

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,279 @@
+## Manage staged changes
+
+The Notary CLI client stages changes before publishing them to the server.
+These chagnes are staged locally in files such that the client can request the latest updates from the notary server, attempt to rebase the staged changes against new updates, and then finally publish if the changes are still valid.
```
spelling nit: "changes"

HuKeping replied:
Does it make sense to add a note for people that it was like a `git rebase`?

riyazdf replied:
@HuKeping great idea!

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+$ notary status <GUN> --reset
+```
+
+Note that `<GUN>` can take on arbitrary values, but when working with Docker Content Trust they are structured like `<url>/<account>/<repository>`.  For example `docker.io/library/alpine` for the [Docker Library Alpine image](https://hub.docker.com/r/library/alpine).
````
nitpick:  I'm not sure, but I think URLs generally include the protocol or scheme (e.g. "https").  Maybe for the GUN, `<url>` should be `domain` or `<registry>`?

HuKeping replied:
Aha, it was `<account>`, we always considered it as `<namespace>` :)

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+In the above command, the `<target_name>` corresponds to the name we want to associate the `<target_file>`'s signature with. Notary will sign the hash of the `<target_file>` into its trusted collection.
+Instead of adding a target by file, you can specify a hash directly:
+```bash
+$ notary addhash -p <GUN> <target_name> -sha256 <sha256Hash>
````
I think this also needs the size flag.

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+
+To remove targets from a trusted collection, you can run:
+```bash
+$ notary remove -p <GUN> <target_name>
````
You can also specify multiple roles to remove a target from - otherwise it defaults to only attempting to remove it from the `targets` role?

cyli replied:
Oh oops, the multi-role stuff is below.  :)  Sorry - it might be good to include `notary remove -p <GUN> <target_name> --roles ...` down there too

riyazdf replied:
Yeah I put the multi-role stuff below because I hadn't introduced delegations yet, I'll add a `notary remove --roles` example too

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,279 @@
+The Notary CLI client generates a new key for the role you specified, and
+prompts you for a passphrase to encrypt it.
+Then you're prompted for the passphrase for the key you're rotating, and if it
+is correct, the Notary CLI client contacts the Notary server to publish the
```
Non-blocking:  wondering if we should call out that Notary CLI auto-publishes immediately, no flag needed?

HuKeping commented on docs/command_reference.md:
```diff
@@ -0,0 +1,279 @@
+3. **Trusted Collection**: A trusted collection is a set of target files of interest to the application.
+4. **Notary Repository**: A notary repository refers to the set of signed metadata files that describe a trusted collection.
+5. **Key roles**:
+    - **Root**: The root role delegates trust to specific keys trusted for all other top-level roles used in the system.
```
Does it make sense to link the top-level to the [relevant part of TUF spec](https://github.com/theupdateframework/tuf/blob/develop/docs/tuf-spec.txt#L246)

riyazdf replied:
I'd rather not link to specific parts of the TUF spec in too many places in the command reference, since I feel like that's more relevant to the other documentation that explains more of the TUF concepts.  Do you think just the general TUF spec link I have below is enough?

HuKeping replied:
agree, two many link doesn't make sense

HuKeping commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+$ alias notary="notary -s <notary_server_url> -d <notary_cache_directory>
+```
+
+When working Docker Content Trust, it is important to connect to Docker Hub's Notary server located at `https://notary.docker.io`, and specify notary's metadata cache as `~/.docker/trust`.
````
The `https:/notary.docker.io` here is not very accurate, if one enabled the DCT and set up his own Notary service, what he need to specify is the URL of his notary server(and it might be the most instance).

riyazdf replied:
I was trying to capture the general case, but I'll make sure to specify this is for images specifically wanting to use the notary server deployed alongside hub.docker.io

HuKeping commented on docs/command_reference.md:
```diff
@@ -0,0 +1,279 @@
+This command will generate targets and snapshot keys locally for the trusted collection, and try to locate a root key to use in the specified metadata cache (from `-d` or the config).
+If notary cannot find a root key, it will generate one.  For all keys, notary will also prompt for a passphrase to encrypt the private key material at rest.
+
+If you'd like to initialize your trusted collection with a specific root key, there is a flag:
```
I think there should be more detail about the customized root key, for example if people use `openssl` to generate a key, that key might not be acceptable by this command.

HuKeping commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+$ notary init <GUN> --rootkey <key_file>
+```
+
+Note that you will have to run a publish after this command, as there are staged changes to initialize the trusted collection that have not been pushed to a notary server.
````
maybe "Note that you will have to run a publish after this command to make it efficient"

HuKeping commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+
+Users can sign content into a notary trusted collection by running:
+```bash
+$ notary add -p <GUN> <target_name> <target_file>
````
Since `-p` is not required, would you mind if we just use `notary add <GUN> <target_name> <target_file>` here?

riyazdf replied:
I wanted to keep `-p` in commands after introducing it in the auto-publish section so that all of the commands will complete the full publish cycle.  I'll make a note in the previous section about how it'll be included going forward for commands that allow it (and that it is optional)

HuKeping replied:
That make sense :+1:

HuKeping commented on docs/command_reference.md:
```diff
@@ -0,0 +1,279 @@
+
+After a rotation, all previously existing keys for the specified role are replaced with the new key.
+
+You can also rotate keys that are stored in the Notary server, such as the keys
```
Non-blocking: "You can .... Notary server with flag `-r`" ?

HuKeping commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+To delegate content signing to other users without sharing the targets key, retrieve a x509 certificate for that user and run:
+
+```bash
+$ notary delegation add -p <GUN> targets/<role> --all-paths user.pem user2.pem
````
I suppose the `--all-paths` should be an option?

riyazdf replied:
I think it's the common use case so I used that, but I'll make a note about it and `--paths`

HuKeping commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+$ notary delegation add -p <GUN> targets/<role> --all-paths user.pem user2.pem
+```
+
+The user can then import the private key for that certificate keypair (using `notary key import`) and use it for signing.
````
Maybe "The delegated user" will be more clear

HuKeping commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+$ notary delegation add -p <GUN> targets/<role> --all-paths user1.pem user2.pem user3.pem
+```
+
+You can also remove keys from a delegation role:
````
Should we also describe the consequence it brings?

The delegation role will not allowed to sign the content any more if I understand right.

HuKeping commented on docs/command_reference.md:
````diff
@@ -0,0 +1,279 @@
+
+We can specify to sign content into specific delegation roles by specifying the `--roles` flag.  This also applies to `notary addhash` and `notary remove`:
+```bash
+$ notary add -p <GUN> <target_name> <target_file> --roles targets/<role>
````
The flag `-p` is not required.

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,299 @@
+
+## Terminology Reference
+1. **GUN**: Notary uses Globally Unique Names (GUNs) to identify trusted collections.
+2. **Target**: Notary refers to the files the framework is used to distribute as "target files".  Target files are opaque to the framework. Whether target files are packages containing multiple files, single text files, or executable binaries is irrelevant to Notary.
```
nitpick: "refers to the files the framework uses"

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,299 @@
+## Terminology Reference
+1. **GUN**: Notary uses Globally Unique Names (GUNs) to identify trusted collections.
+2. **Target**: Notary refers to the files the framework is used to distribute as "target files".  Target files are opaque to the framework. Whether target files are packages containing multiple files, single text files, or executable binaries is irrelevant to Notary.
+3. **Trusted Collection**: A trusted collection is a set of target files of interest to the application.
```
nitpick:  By application do you mean application other than notary, or notary itself?  (it might be good to clarify this)

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,299 @@
+$ notary init <GUN> --rootkey <key_file>
+```
+
+Note that you will have to run a publish after this command for it to take effect, as there are staged changes to initialize the trusted collection that have not been pushed to a notary server.
````
nitpick:  "as it creates staged changes..." might be clearer about where those staged changes came from.

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,299 @@
+$ notary status <GUN> --reset
+```
+
+Note that `<GUN>` can take on arbitrary values, but when working with Docker Content Trust they are structured like `<domain>/<account>/<repository>`.  For example `docker.io/library/alpine` for the [Docker Library Alpine image](https://hub.docker.com/r/library/alpine).
````
Non-blocking:  should the description of the GUN be moved to the "terminology section", or alternately the `notary init` section where it is next mentioned?

riyazdf replied:
Good idea, I'll move it to terminology

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,299 @@
+$ notary add -p <GUN> <target_name> <target_file>
+```
+
+In the above command, the `<target_name>` corresponds to the name we want to associate the `<target_file>`'s signature with. Notary will sign the hash of the `<target_file>` into its trusted collection.
````
Non-blocking:  should we use the word "signature" here, since the rest of notary is about signing things?  Since we're not specifically signing this `<target_file>` itself, maybe something like "`<target_name>` corresponds to the name we want `<target_file>` to be known as in the trusted collection"?

riyazdf replied:
Good point, I'll update this

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,299 @@
+
+## Managing targets in delegation roles
+
+We can specify to sign content into specific delegation roles by specifying the `--roles` flag.  This also applies to `notary addhash` and `notary remove`.
```
nitpick: maybe "we can specify which delegation roles to sign content into by using the `--roles` flag"?

cyli commented on docs/command_reference.md:
```diff
@@ -0,0 +1,299 @@
+$ notary addhash -p <GUN> <target_name> <byte_size> --sha256 <sha256Hash> --roles targets/<role>
+
+# Remove content from a specific delegation role
+$ notary add -p <GUN> <target_name> <target_file> --roles targets/<role>
```
`notary remove`

cyli commented:
Thanks for working on this @riyazdf!  This will be a super helpful reference  :)  I have a couple of nitpicks above, but other than that, LGTM!

cyli commented on docs/command_reference.md:
````diff
@@ -0,0 +1,297 @@
+$ notary add -p <GUN> <target_name> <target_file>
+```
+
+In the above command, the `<target_name>` corresponds to the name we want to associate the `<target_file>` known as in the trusted collection. Notary will sign the hash of the `<target_file>` into its trusted collection.
````
nitpick: "the name we want `<target_file>` to be known as in the trusted collection", or alternately "the name we want to associate the `<target file>` with in the trusted collection"  :)

endophage commented on docs/command_reference.md:
```diff
@@ -0,0 +1,297 @@
+
+## Terminology Reference
+1. **GUN**: Notary uses Globally Unique Names (GUNs) to identify trusted collections.  Note that `<GUN>` can take on arbitrary values, but when working with Docker Content Trust they are structured like `<domain>/<account>/<repository>`.  For example `docker.io/library/alpine` for the [Docker Library Alpine image](https://hub.docker.com/r/library/alpine).
+2. **Target**: Notary refers to the files the framework uses to distribute as "target files".  Target files are opaque to the framework. Whether target files are packages containing multiple files, single text files, or executable binaries is irrelevant to Notary.
```
First sentence reads poorly.

endophage commented on docs/command_reference.md:
````diff
@@ -0,0 +1,297 @@
+$ notary key import <pemfile> --role <key_role> --gun <key_gun>
+```
+
+The `--role` and `--gun` flags can be provided to specify a role and GUN to import the key to, which overrides any information contained in PEM headers.
````
I don't think we implemented it as override. They fill in blank data but don't override existing data afaik.

riyazdf replied:
oops you're right, I must've been thinking of something else 👍

endophage commented on docs/command_reference.md:
```diff
@@ -0,0 +1,297 @@
+
+The `--role` and `--gun` flags can be provided to specify a role and GUN to import the key to, which overrides any information contained in PEM headers.
+Note that for root and delegation keys, the `--gun` flag is ignored because these keys can be shared across GUNs.
+If no `--role` or `--gun` is given, notary will assume that the key is to be used for a delegation role.
```
Might want to note that it will be available for signing regardless of role, but will be shown as a delegation key in commands such as `notary key list`

endophage commented on docs/command_reference.md:
```diff
@@ -0,0 +1,297 @@
+# Remove the given keys from a delegation role
+$ notary delegation remove -p <GUN> targets/<role> <keyID1> <keyID2>
+
+# Alternatively, you can remove keys from all delegation roles
```
Include the use case in which you might want to use this option: delegation key compromise.

endophage commented on docs/command_reference.md:
````diff
@@ -0,0 +1,297 @@
+* fatal: Get <URL>/v2/: x509: certificate signed by unknown authority
+```
+The Notary CLI must be configured to trust the root certificate authority of the server it is communicating with.
+This can be configured by specifying the root certificate with the `--tlscacert` flag or by editing the Notary client configuration file.
````
or by adding the CA to your system CAs

endophage commented:
LGTM. Merge on green!
